pkg/dynamicinstrumentation/testutil/e2e_test.go

```diff
@@ -0,0 +1,188 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf && arm64
+
+package testutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"reflect"
+	"strings"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation"
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/diconfig"
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/features"
+	"github.com/cilium/ebpf/rlimit"
+	"github.com/kr/pretty"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestGoDI(t *testing.T) {
+	if err := rlimit.RemoveMemlock(); err != nil {
+		t.Error(err)
+	}
+
+	if features.HaveMapType(ebpf.RingBuf) != nil {
+		t.Skip("ringbuffers not supported on this kernel")
+	}
+
+	sampleServicePath := BuildSampleService(t)
+	cmd := exec.Command(sampleServicePath)
+	cmd.Env = []string{
+		"DD_DYNAMIC_INSTRUMENTATION_ENABLED=true",
+		"DD_SERVICE=go-di-sample-service",
+		"DD_DYNAMIC_INSTRUMENTATION_OFFLINE=true",
+	}
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	require.NoError(t, cmd.Start())
+	t.Cleanup(func() {
+		log.Println(cmd.Process.Kill())
+	})
+
+	eventOutputWriter := &eventOutputTestWriter{
+		t: t,
+	}
+
+	opts := &dynamicinstrumentation.DIOptions{
+		RateLimitPerProbePerSecond: 0.0,
+		ReaderWriterOptions: dynamicinstrumentation.ReaderWriterOptions{
+			CustomReaderWriters: true,
+			SnapshotWriter:      eventOutputWriter,
+			DiagnosticWriter:    os.Stderr,
+		},
+	}
+
+	var (
+		GoDI *dynamicinstrumentation.GoDI
+		err  error
+	)
+
+	GoDI, err = dynamicinstrumentation.RunDynamicInstrumentation(opts)
+	require.NoError(t, err)
+
+	// Give DI time to start up
+	time.Sleep(time.Second * 5)
+
+	cm, ok := GoDI.ConfigManager.(*diconfig.ReaderConfigManager)
+	if !ok {
+		t.Error("Config manager is of wrong type")
+	}
+
+	cfgTemplate, err := template.New("config_template").Parse(configTemplateText)
+	require.NoError(t, err)
+
+	b := []byte{}
+	var buf *bytes.Buffer
+	for function, expectedCaptureValue := range expectedCaptures {
+		// Generate config for this function
+		buf = bytes.NewBuffer(b)
+		functionWithoutPackagePrefix, _ := strings.CutPrefix(function, "main.")
+		err = cfgTemplate.Execute(buf, configDataType{functionWithoutPackagePrefix})
+		require.NoError(t, err)
+		eventOutputWriter.doCompare = false
+		eventOutputWriter.expectedResult = expectedCaptureValue
+
+		// Read the configuration via the config manager
+		_, err := cm.ConfigReader.Read(buf.Bytes())
```

Comment:
If the type assertion above (line 81) fails, then this will be `nil`. Should the type assertion cause the test to stop? If so, use `t.Fatal` instead of `t.Error`.